### PR TITLE
Rework manual Connection close in C++

### PR DESCRIPTION
**Pull request by bernardnormier**
This is a companion PR for #2651, for C++ and derived language.

In C++ and derived languages, unlike C#, the new async close is called simply "close". 

The Async suffix is a C# convention, not a general convention. We use this convention in C++ when we provide both sync and async overloads (e.g. for proxy member functions) or for historical reasons (the generated AMD methods), but we should not use this convention when there is only one API that happens to be async (e.g. Object::dispatch, Connection::close).

I anticipate this rule (no Async-suffix) will be helpful when we refactor other APIs like shutdown.

PHP, Ruby: close is currently sync, i.e. when you all `connection.close()`, the call waits until the connection is closed. It also throws an exception if the closure is not graceful.

Python: `connection.close()` is for now synchronous, until we figure out a pattern for async APIs in Python

MATLAB and Swift: `close()` is asynchronous,

### CODE REVIEW

File: cpp/include/Ice/Connection.h [hunk 1]
```diff
@@ -144,11 +123,26 @@ namespace Ice
...
+         * CloseTimeoutException.
+         */
+        virtual void close(std::function<void(std::exception_ptr)> whenClosed) noexcept = 0;
+
+        /**
+         * Starts a graceful closure of this connection once all outstanding invocations have completed.
+         * @return A future that completes then the connection is closed.
          */
-        virtual void close(ConnectionClose mode) noexcept = 0;
+        std::future<void> close() noexcept;
```
Comment (bernardnormier) [inline cpp/include/Ice/Connection.h:145]: The callback API is not really practical to use. That's why I provided this future overload.

File: cpp/src/Ice/Connection.cpp [hunk 2]
```diff
@@ -32,3 +32,39 @@ Ice::Connection::flushBatchRequestsAsync(CompressBatch compress)
         [promise](bool) { promise->set_value(); });
     return promise->get_future();
 }
+
+future<void>
+Ice::Connection::close() noexcept
+{
+    auto sharedPromise = make_shared<promise<void>>();
```
Comment (bernardnormier) [inline cpp/src/Ice/Connection.cpp:39]: Let me know if you see a nicer pattern for this callback/future adaptation.

File: cpp/src/Ice/Connection.cpp [hunk 3]
```diff
@@ -32,3 +32,39 @@ Ice::Connection::flushBatchRequestsAsync(CompressBatch compress)
...
+    close(
+        [sharedPromise](exception_ptr closeException)
+        {
+            try
+            {
+                rethrow_exception(closeException);
+            }
+            catch (const ConnectionClosedException&)
```
Comment (bernardnormier) [inline cpp/src/Ice/Connection.cpp:47]: We need to keep this list of "graceful close" exceptions in sync in all languages.

Comment (bernardnormier) [reply cpp/src/Ice/Connection.cpp]: Removed.

**Review by InsertCreativityHere — APPROVED** [review]
Looks good!

**Pull request merged**